source/nvwave.py

```diff
@@ -189,48 +270,93 @@ def open(self):
+					winmm.waveOutOpen(
+						byref(waveout),
+						self._outputDeviceID,
+						LPWAVEFORMATEX(wfx),
+						self._waveout_event,
+						0,
+						CALLBACK_EVENT
+					)
+			except WindowsError:
+				log.debug(
+					f"Error opening"
+					f" outputDeviceName: {self._outputDeviceName}"
+					f" with id: {self._outputDeviceID}"
+				)
+				if self._outputDeviceID != WAVE_MAPPER:
+					log.debug(f"Falling back to WAVE_MAPPER")
+					self._setCurrentDevice(WAVE_MAPPER)
+					self.open()
+				else:
+					log.warning(f"Unable to open WAVE_MAPPER device, there may be no audio devices.")
+					raise  # can't open the default device.
+				return
 			self._waveout = waveout.value
 			self._prev_whdr = None
 
-	def feed(self, data, onDone=None):
+	def feed(
+			self,
+			data: bytes,
+			onDone: typing.Optional[typing.Callable] = None
+	) -> None:
 		"""Feed a chunk of audio data to be played.
 		This is normally synchronous.
-		However, synchronisation occurs on the previous chunk, rather than the current chunk; i.e. calling this while no audio is playing will begin playing the chunk but return immediately.
-		This allows for uninterrupted playback as long as a new chunk is fed before the previous chunk has finished playing.
+		However, synchronisation occurs on the previous chunk, rather than the current chunk;
+		i.e. calling this while no audio is playing will begin playing the chunk
+		but return immediately.
+		This allows for uninterrupted playback as long as a new chunk is fed before
+		the previous chunk has finished playing.
 		@param data: Waveform audio in the format specified when this instance was constructed.
-		@type data: str
 		@param onDone: Function to call when this chunk has finished playing.
-		@type onDone: callable
 		@raise WindowsError: If there was an error playing the audio.
 		"""
 		if not self._minBufferSize:
-			return self._feedUnbuffered(data, onDone=onDone)
+			self._feedUnbuffered_handleErrors(data, onDone=onDone)
+			return
 		self._buffer += data
 		# If onDone was specified, we must play audio regardless of the minimum buffer size
 		# so we can accurately call onDone at the end of this chunk.
 		if onDone or len(self._buffer) > self._minBufferSize:
-			self._feedUnbuffered(self._buffer, onDone=onDone)
+			data = self._buffer
 			self._buffer = b""
+			self._feedUnbuffered_handleErrors(data, onDone=onDone)
+
+	def _feedUnbuffered_handleErrors(self, data, onDone=None) -> bool:
+		"""Tries to feed the device, on error resets the device and tries again.
+		@return: False if second attempt fails
+		"""
+		try:
+			self._feedUnbuffered(data, onDone=onDone)
+			return True
+		except WindowsError:
+			log.warning("Error during feed. Resetting the device.")
+			try:
+				self._close()  # don't try to call stop on a "broken" device.
+				self._setCurrentDevice(self._preferredDeviceName)
+				self.open()
+				self._feedUnbuffered(data, onDone=onDone)
+			except Exception:
+				log.debugWarning("Unable to recover.")

```

Comment:
Can this please be log.exception rather than log.debugWarning.
This allows us to see the traceback, which a part from showing the true error, gives us more context, as log calls do not log the source code line number by the way.